source/extensions/filters/listener/tls_inspector/tls_inspector.cc

```diff
@@ -152,39 +112,31 @@ void Filter::onServername(absl::string_view name) {
   clienthello_success_ = true;
 }
 
-ParseState Filter::onRead() {
-  // This receive code is somewhat complicated, because it must be done as a MSG_PEEK because
-  // there is no way for a listener-filter to pass payload data to the ConnectionImpl and filters
-  // that get created later.
-  //
-  // We request from the file descriptor to get events every time new data is available,
-  // even if previous data has not been read, which is always the case due to MSG_PEEK. When
-  // the TlsInspector completes and passes the socket along, a new FileEvent is created for the
-  // socket, so that new event is immediately signaled as readable because it is new and the socket
-  // is readable, even though no new events have occurred.
-  //
-  // TODO(ggreenway): write an integration test to ensure the events work as expected on all
-  // platforms.
-  const auto result = cb_->socket().ioHandle().recv(buf_, config_->maxClientHelloSize(), MSG_PEEK);
-  ENVOY_LOG(trace, "tls inspector: recv: {}", result.return_value_);
-
-  if (!result.ok()) {
-    if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
-      return ParseState::Continue;
-    }
-    config_->stats().read_error_.inc();
-    return ParseState::Error;
-  }
+Network::FilterStatus Filter::onData(Network::ListenerFilterBuffer& buffer) {
+  auto rc = buffer.copyOut(*buffer_, config_->maxClientHelloSize());
+  ENVOY_LOG(trace, "tls inspector: recv: {}", rc);
 
   // Because we're doing a MSG_PEEK, data we've seen before gets returned every time, so
   // skip over what we've already processed.
-  if (static_cast<uint64_t>(result.return_value_) > read_) {
-    const uint8_t* data = buf_ + read_;
-    const size_t len = result.return_value_ - read_;
-    read_ = result.return_value_;
-    return parseClientHello(data, len);
+  if (static_cast<uint64_t>(rc) > read_) {
+    uint8_t* buf = static_cast<uint8_t*>(buffer_->linearize(buffer_->length()));
```

Comment:
@mattklein123 if I use `copyOut(Buffer::Instance, uint64_t length)`, then the filter need to use the Buffer::Instance, but the filter also depends on the raw buffer(also for proxy filter), so I didn't see too much benefit, and have to `linearize` the Buffer to get the raw buffer.

Or we can just using the Buffer::Instance inside the ListenerFilterBufferImpl. keep the using raw buffer for the parameter of copyOut `copyOut(void* buffer, uint64_t length)`. You can reference the commit https://github.com/envoyproxy/envoy/pull/17395/commits/1fc908d7307a3cb025f8e95dd163ccdf0df2ac1a

Let me know which way is your preference.

Thanks for the review again!

Reply:
Yeah I agree per my other comments. We can go back.

Reply:
Buffer::RawSlice will allow the filter to modify the data through the RawSlice::mem_ directly. Per the discussion we have, we want a const Buffer.

Reply:
Maybe make `ConstRawSlice` that has a const pointer then. Or use any other convenient type that is a pointer+length with a const pointer.

Reply:
oh, right, that works also. thanks! let me update the PR.